sound/soc/sof/ipc4-pcm.c

```diff
@@ -32,65 +60,184 @@ int sof_ipc4_set_pipeline_state(struct snd_sof_dev *sdev, u32 id, u32 state)
+		/*
+		 * IPC4 requires pipelines to be triggered in order starting at the sink and
+		 * walking all the way to the source. So the starting widget for playback is of
+		 * DAI type and for capture, it is of AIF type.
+		 */
+		if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK && !WIDGET_IS_DAI(widget->id))
+			continue;
 
-		/* skip triggering backend pipelines */
-		if (pipeline->is_backend || pipeline->state == state)
+		if (substream->stream == SNDRV_PCM_STREAM_CAPTURE && !WIDGET_IS_AIF(widget->id))
 			continue;
 
+		ret = sof_ipc4_get_pipelines_in_order(sdev, widget, list, &pipeline_info);
+		if (ret < 0)
+			return ret;
+	}
+
+	if (!pipeline_info.count)
+		return ret;
+
+	/* trigger a single pipeline if count is 1 */
+	if (pipeline_info.count == 1) {
+		pipeline_widget = pipeline_info.pipelines[0];
+		pipeline = pipeline_widget->private;
+
 		/* first set the pipeline to PAUSED state */
 		if (pipeline->state != SOF_IPC4_PIPE_PAUSED) {
 			ret = sof_ipc4_set_pipeline_state(sdev, pipeline_widget->instance_id,
 							  SOF_IPC4_PIPE_PAUSED);
+
 			if (ret < 0) {
 				dev_err(sdev->dev, "failed to pause pipeline %d\n",
-					swidget->pipeline_id);
+					pipeline_widget->pipeline_id);
 				return ret;
 			}
 		}
 
 		pipeline->state = SOF_IPC4_PIPE_PAUSED;
 
 		if (pipeline->state == state)
-			continue;
+			return ret;
+
 
 		/* then set the final state */
 		ret = sof_ipc4_set_pipeline_state(sdev, pipeline_widget->instance_id, state);
 		if (ret < 0) {
-			dev_err(sdev->dev, "failed to set state %d for pipeline %d\n",
-				state, swidget->pipeline_id);
-			break;
+			dev_err(sdev->dev, "failed to set state %d pipeline %d\n",
+				state, pipeline_widget->pipeline_id);
+			return ret;
 		}
 
 		pipeline->state = state;
+		return ret;
+	}
+
+	/*
+	 * check if the pipelines are paused already.
+	 * It's enough to check the first pipeline's state
+	 */
+	pipeline_widget = pipeline_info.pipelines[0];
+	pipeline = pipeline_widget->private;
+	if (pipeline->state == SOF_IPC4_PIPE_PAUSED)
+		goto out;
+
+	/* pause all pipelines */
+	ret = sof_ipc4_set_multi_pipeline_state(sdev, SOF_IPC4_PIPE_PAUSED, &pipeline_info);
+	if (ret < 0) {
+		dev_err(sdev->dev, "failed to pause all pipelines\n");
+		return ret;
+	}
+
+	for (i = 0; i < pipeline_info.count; i++) {
+		pipeline_widget = pipeline_info.pipelines[i];
+		pipeline = pipeline_widget->private;
+		pipeline->state = SOF_IPC4_PIPE_PAUSED;
+	}
+
+out:
+	/* check if the pipelines are in the final state already */
+	pipeline_widget = pipeline_info.pipelines[0];
+	pipeline = pipeline_widget->private;
+	if (pipeline->state == state)
+		return ret;
+
+	ret = sof_ipc4_set_multi_pipeline_state(sdev, state, &pipeline_info);
+	if (ret < 0) {
+		dev_err(sdev->dev, "failed to set final state %d for all pipelines\n", state);
+		return ret;
+	}
+
+	for (i = 0; i < pipeline_info.count; i++) {
+		pipeline_widget = pipeline_info.pipelines[i];
+		pipeline = pipeline_widget->private;
+		pipeline->state = state;
```

Comment:
So, we still have the separate BE (DAI) pipeline trigger and we handle the non BE (DAI) pipelines in batch mode?

Reply:
no if there are multiple BE DAI's, their pipelines will get triggered in the respective BE trigger callbacks. There wouldnt be a batch mode there

Reply:
In here it does not matter if we have triggered it or not like in case of the pause?

Reply:
it does matter and if it is triggered it will be in the right state already no?

Reply:
I meant the ones newly triggered will need an update to the state and the others will be the same state already. So, no need to check here